File: 0322-coin-change/0322-coin-change.cpp
class Solution {
public:
    int coinChange(vector<int>& coins, int amount) {
        
        unordered_set<int> prev;
        prev.insert(amount);
        
        int count = 0;
        unordered_set<int> next;
        
        while (!prev.empty() && prev.find(0) == prev.end()) {
            for (int v : prev) {
                for (int c : coins) {
                    if (v-c >= 0 && prev.find(v-c) == prev.end()) {
                        next.insert(v-c);
                    }
                }
            }
            prev = next;
            next.clear();
            count++;
        }
        
        if (prev.empty()) {
            return -1;
        }
        
        return count;
        
    }
};

File: 0322-coin-change/0322-coin-change.py
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        memo = {}
        
        def recurse(m) -> int:
            if m == 0:
                return 0
            elif m in coins:
                return 1
            elif m < 0:
                return float('inf')
            
            if m in memo.keys():
                return memo[m]
            
            # try each coin
            min_count = float('inf')
            for c in coins:
                memo[m-c] = recurse(m-c)
                if memo[m-c] < min_count:
                    min_count = memo[m-c]
            if min_count == float('inf'):
                return float('inf')
            return min_count + 1
        
        ans = recurse(amount)
        return ans if ans != float('inf') else -1